src/compiler.ts

```diff
@@ -6735,23 +6772,33 @@ export class Compiler extends DiagnosticEmitter {
           }
           // If we're calling a local we know to be a closure, then we must still be in the creator functions
           // scope. Because of this, we should update the values of locals that are still available
-          return module.block(null, [
-            this.injectClosedLocals(local),
-            this.compileCallIndirect(
-              assert(signature),
-              module.load(
-                local.type.byteSize,
-                local.type.is(TypeFlags.SIGNED),
-                module.local_get(local.index, nativeSizeType),
-                this.options.nativeSizeType,
-                0
-              ),
-              expression.args,
-              expression,
+          var nativeContextualType = contextualType.toNativeType();
+          var immediatelyDropped = contextualType == Type.void;
+          var exprs = Array<ExpressionRef>(immediatelyDropped ? 3 : 4)
+
+          exprs[0] = this.injectClosedLocals(local);
+          exprs[1] = this.compileCallIndirect(
+            assert(signature),
+            module.load(
+              local.type.byteSize,
+              local.type.is(TypeFlags.SIGNED),
               module.local_get(local.index, nativeSizeType),
-              contextualType == Type.void
-            )
-          ], signature.returnType.toNativeType());
+              this.options.nativeSizeType,
+              0
+            ),
+            expression.args,
+            expression,
+            module.local_get(local.index, nativeSizeType),
+            immediatelyDropped
+          );
+          exprs[2] = this.writebackClosedLocals(local)
+
+          if(!immediatelyDropped) {
+            let result_local = this.currentFlow.getTempLocal(contextualType);
+            exprs[1] = module.local_set(result_local.index, exprs[1]);
+            exprs[3] = module.local_get(result_local.index, nativeContextualType);
+          }
+          return module.block(null, exprs, nativeContextualType);
```

Comment:
This is what's called when we call a closure while it's still in the same scope as what it's closed over.  The reason this got so much uglier is because doing something after a function call is not great- Binaryen yells at you because 'writebackClosedLocals' returns `none`, even though the result should be on the stack from the `compileCallIndirect`.

Maybe there's a cleaner way to do this other than assigning the return value to a local? Not sure